Fb jpms - fixes for illegal access to private information in java.* packages.

struberg opened the pull request:
please review and give feedback. Can push it later myself or you just apply it - txs!

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/CompareToBuilder.java:
```diff
@@ -115,6 +116,11 @@ private static void reflectionAppend(
+            builder.append(lhs, rhs);
+            return;
+        }
+
```
Extra blank lines not needed IMO.

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/CompareToBuilder.java:
```diff
@@ -115,6 +116,11 @@ private static void reflectionAppend(
         final boolean useTransients,
         final String[] excludeFields) {
 
+        if ((lhs instanceof CharSequence || lhs instanceof Number || lhs instanceof Temporal) && lhs instanceof Comparable) {
```
Needs a // comment IMO

struberg replied:
resolved

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/EqualsBuilder.java:
```diff
@@ -731,7 +732,8 @@ public EqualsBuilder append(final Object lhs, final Object rhs) {
             appendArray(lhs, rhs);
         } else // The simple case, not an array, just test the element
-        if (testRecursive && !ClassUtils.isPrimitiveOrWrapper(lhsClass)) {
+        if (testRecursive && !ClassUtils.isPrimitiveOrWrapper(lhsClass)
```
Needs a // comment IMO

struberg replied:
is now in the JavaDocs of Reflection#isJavaInternalClass(). That should resolve this and all similar situations.

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/EqualsBuilder.java:
```diff
@@ -956,7 +958,8 @@ public EqualsBuilder reflectionAppend(final Object lhs, final Object rhs) {
 
         try {
-            if (testClass.isArray()) {
+            if (testClass.isArray() || Number.class.isAssignableFrom(testClass)
```
Needs a // comment IMO

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/HashCodeBuilder.java:
```diff
@@ -180,6 +181,10 @@ private static void reflectionAppend(final Object object, final Class<?> clazz,
         if (isRegistered(object)) {
             return;
         }
+        if (object instanceof CharSequence || object instanceof Number || object instanceof Temporal) {
```
Needs a // comment IMO. Why not refactor the test exporession into a common (if possible package-private) method with the same code above?

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:
```diff
@@ -867,4 +874,21 @@ private void validate() {
+            getStringBuffer().append("value=").append(getObject().toString());
+            return true;
+        }
+        
```
Extra blank line not needed IMO. Want I want call attention to something, I write a // comment.

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:
```diff
@@ -867,4 +874,21 @@ private void validate() {
+            return true;
+        }
+        
+        if (value.getClass().isArray() || value instanceof Collection || value instanceof Map) {
```
Can `value` be null here?

struberg replied:
nope, that is covered in line 846 of toString()

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:
```diff
@@ -867,4 +874,21 @@ private void validate() {
+            getStyle().append(getStringBuffer(), null, value, true);
+            return true;
+        }
+        
```
Extra empty line not needed IMO.  Want I want call attention to something, I write a // comment.

garydgregory reviewed (changes requested):
Hi @struberg 
I have a few comments scattered throughout.

ppkarwasz commented on pom.xml:
```diff
@@ -202,6 +202,7 @@
               <exclude>src/site/resources/download_lang.cgi</exclude>
               <exclude>src/site/resources/release-notes/RELEASE-NOTES-*.txt</exclude>
               <exclude>src/test/resources/lang-708-input.txt</exclude>
+              <exclude>nb-configuration.xml</exclude>
```
This is not necessary, since RAT has a implicit exclusion list populated from the `.gitignore` file.

struberg replied:
that file is generated by NetBeans when using it. So no, does not have an AL header.
I've added the file to .gitignore but checkstyle did still complain.

ppkarwasz replied:
Did Checkstyle or Apache RAT fail? Apache RAT should be OK without this line, Checkstyle does not read the configuration of Apache RAT AFAIK.

struberg replied:
I'll try again without

struberg replied:
yup, you're right it seems. Works now.

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:
```diff
@@ -867,4 +874,21 @@ private void validate() {
+    private boolean handleNativeClasses() {
+        Object value = getObject();
+        if (value instanceof Number || value instanceof Boolean || value instanceof Character
+                || value instanceof TemporalAccessor) {
```
This expression uses `TemporalAccessor` here but the new `isJavaInternalClass()` uses `Temporal`. `Temporal` extends `TemporalAccessor`, why is there a difference? It seems this expression could be rewritten to call  `isJavaInternalClass()`, this centralize the types to check.

garydgregory reviewed (changes requested):
Hi @struberg 

See my comments and @ppkarwasz's.

struberg commented on src/main/java/org/apache/commons/lang3/builder/ReflectionToStringBuilder.java:
```diff
@@ -867,4 +874,21 @@ private void validate() {
+    private boolean handleNativeClasses() {
+        Object value = getObject();
+        if (value instanceof Number || value instanceof Boolean || value instanceof Character
+            || value instanceof TemporalAccessor ) {
```
good catch, moved all to TomporalAccessor in Reflection.isJavaInternalClass()

garydgregory reviewed (changes requested):
Hi @struberg 

We cannot test that an object implements a stock interface like `TemporalAccessor` or `CharSequence` because this won't work with custom implementations (`TemporalAmount` is also not accounted for, and others I am sure).

See the new test I added in git master: `EqualsBuilderReflectJreImplementationTest`

I think we can only test for an actual class or the the package that class lives in. 

TY

struberg commented on src/test/java/org/apache/commons/lang3/builder/ToStringBuilderTest.java:
```diff
@@ -229,21 +230,33 @@ public void assertReflectionArray(final String expected, final Object actual) {
+        int y = 2;
+    }
+
+    private static class HirCFixture {
```
Hierarchic

struberg replied:
Regarding `TemporalAccessor` and `CharSequence`: happy to replace CharSequence with a String.class check. Not quite sure about TemporalAccessor, no strong feeling, we could also move to `Temporal` if you want.

garydgregory replied:
That's not what I'm saying ;-) What I'm trying to say is that the check _cannot_ use _any_ interfaces.

struberg commented on src/test/java/org/apache/commons/lang3/builder/ToStringBuilderTest.java:
```diff
@@ -229,21 +230,33 @@ public void assertReflectionArray(final String expected, final Object actual) {
-        final Integer val = Integer.valueOf(5);
+        final HirAFixture val = new HirAFixture();
         final ReflectionToStringBuilder test = new ReflectionToStringBuilder(val);
-        test.setUpToClass(Number.class);
```
In this very case this test simply makes no sense. setUpToClass is INCLUDING the very class given. But if you introspect into any java.lang.Number, then it will just blow up your whole application from Java9 onwards...

garydgregory reviewed (dismissed):
Hello @struberg 

Checking against an interface in `isJavaInternalClass()` is not going to work. 

I updated `EqualsBuilderReflectJreImplementationTest` in git master.

struberg commented:
> Hello @struberg
> 
> Checking against an interface in `isJavaInternalClass()` is not going to work.
> 
> I updated `EqualsBuilderReflectJreImplementationTest` in git master.

Gary, this has already long been resolved in https://github.com/apache/commons-lang/pull/1328/commits/f20639a039476826b9c146d96f64e6979ea7d772 and https://github.com/apache/commons-lang/pull/1328/commits/ec4150bc13cfd2705e5d672f47706930af578c39

I've even merged your changes (and removed that new --add-opens again...) from main to this very branch.

garydgregory commented:
> > Hello @struberg
> > Checking against an interface in `isJavaInternalClass()` is not going to work.
> > I updated `EqualsBuilderReflectJreImplementationTest` in git master.
> 
> Gary, this has already long been resolved in [f20639a](https://github.com/apache/commons-lang/commit/f20639a039476826b9c146d96f64e6979ea7d772) and [ec4150b](https://github.com/apache/commons-lang/commit/ec4150bc13cfd2705e5d672f47706930af578c39)
> 
> I've even merged your changes (and removed that new --add-opens again...) from main to this very branch.
> 
> This branch clearly improves the situation with standard jpms setup for a lot of common cases. Of course, this does not yet fix ALL the problems (albeit it's a good start). But if people come across another `InaccessibleObjectException` then they can still --add-opens as needed. So the current version is clearly an improvement. If you agree, then we should merge it into main instead of yet more parallel work in different branches...

Hello @struberg 

The issue is not that this PR is a wants to start improving something, the issue is that this PR _breaks_ existing behavior.

You cannot check for interfaces to short-circuit logic in `Reflection.isJavaInternalClass()`. This will break existing applications.

The branch is not up to date with master. This fails:
```
git clone https://github.com/struberg/commons-lang.git 
cd commons-lang
git checkout fb_jpms
git remote add upstream https://github.com/apache/commons-lang.git
git pull upstream master
mvn clean test -Dtest=EqualsBuilderReflectJreImplementationTest
```

struberg commented:
> You cannot check for interfaces to short-circuit logic in `Reflection.isJavaInternalClass()`. This will break existing applications.

Which of those is an interface?
https://github.com/struberg/commons-lang/blob/ec4150bc13cfd2705e5d672f47706930af578c39/src/main/java/org/apache/commons/lang3/builder/Reflection.java#L62

The thing you are talking about IS ALREADY RESOLVED since yesterday.

And one other thing: commons-lang3 ReflectionToString ReflectionEquals etc was BROKEN on anything Java9++ - period!
The build is ONLY green because you just disabled all the modularity - which is something not everyone can do in real projects!

My spare time to work on this stuff is very limited - as is yours I bet. So please use our both time productively. As I said: not everything will fully work from the beginning. You just added `java.time.Period`, fine, good catch. But why on earth don't we just take that PR and commit it and then improve it TOGETHER? I'm not willing to do more empty rounds getting sticks thrown at me at will. I really appreciate the review feedback as it indeed did highly improve the code. But now is the point to decide whether the direction is right or not - and then improve things from here TOGETHER.

struberg commented:
Btw, I'm looking for another method name for `isJavaInternalClass()`. It should rather express that the given clazz cannot be introspected via reflection. Maybe something like `isNonIntrospectibleClass()`?

Further down the road we need some way to tell whether a class is open or not. And all that in a way which also works on Java8 still...

struberg commented:
> I guess you could be bothered to look it up

Indeed, had in the back of my mind that Temporal is an abstract class.
Latest version uses reflection to figure that a class is not accessible. This works, but maybe we should try a mixture of the old approach + the new one?

Thought about having a static `Map<Class, Boolean> KNOWN_INACCESSIBLE_CLASSES` which contains the concrete classes like LocalDate, Boolean, Character, Integer, etc.
And then in a second step we do the reflection check. But as this might also have unexpected results (people might forget that they closed the jars) we should probably log it out?

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/Reflection.java:
```diff
@@ -18,13 +18,62 @@
+                Float.class,
+                Double.class,
+                java.util.Date.class,
+                java.sql.Date.class,
```
This is not specific to this PR: I don't think we should introduce a _new JPMS module dependency_, here on `java.sql`. Lang does not use any classes from `java.sql` so this would drag in a new JPMS module. For 4.0, we've mentioned dropping our dependency on `java.desktop` and only depending on `java.base`.

struberg replied:
yup true

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/Reflection.java:
```diff
@@ -41,4 +90,44 @@ static Object getUnchecked(final Field field, final Object obj) {
+                f.setAccessible(true);
+            }
+
+            nonIntrospectibleClasses.put(clazz, false);
```
This looks problematic since the map will grow uncontrollably, possibly until every class on your classpath gets referenced, there's no way to know really.

struberg replied:
this is a trade-off vs performance. it will only cache those classes which got hit in a reflection equals, toString, etc. Thus in practice it is a very limited set of classes I'd say. Can you think of a possible place where we can hold such a cache at least during a single invocation? Maybe inside the builder and handing the Map as another parameter to this method?

garydgregory commented on src/main/java/org/apache/commons/lang3/builder/Reflection.java:
```diff
@@ -41,4 +90,44 @@ static Object getUnchecked(final Field field, final Object obj) {
+    }
+
+    /**
+     * Check whether the given class internas can be accessed. This might return false in Java9
```
"internas" spelling

garydgregory reviewed (changes requested):
Hi All,

So far, I'm still against this change.

These features are _designed to use reflection_, and this PR tries to turn some of that off. 

It seems to me like it will cause unintended consequences beyond all of the ones I pointed out through the new unit tests I wrote which showed the initial implementation would break existing apps. 

Now, we have a different implementation, which brings in a new set of problems: 
- It depends on a new JPMS module `java.sql`
- it grows a map without any bounds (a CVE waiting to happen IMO)
- you can't turn it off
- The behavior is completely hidden from the user

ppkarwasz commented:
> So far, I'm still against this change.
> 
> These features are _designed to use reflection_, and this PR tries to turn some of that off.

I agree. All the `reflectionX` methods have a warning:

> Fields can be private, thus `AccessibleObject.setAccessible` is used to bypass normal access control checks. This will fail under a security manager unless the appropriate permissions are set.

The warning might require an update, but it seems clear to me that it also applies to the JPMS access checks. So it is up to caller to ensure that the caller to ensure that the class is open to `org.apache.commons.lang3`. If the caller uses `testRecursive`, he must also ensure that the appropriate permissions are set for all the field types.

garydgregory commented:
This discussion is mirrored on the apache list here: 
https://lists.apache.org/thread/gwljc3ylz3cjvbg9m7sv9dtj139obprz

garydgregory reviewed (dismissed):
Note that this PR is worse than I thought because we already have a feature to bypass reflection! Use `EqualsBuilder.setBypassReflectionClasses(List<Class<?>>)`.

struberg commented:
further comments over at the commons dev list.

struberg commented:
> I agree. All the `reflectionX` methods have a warning:

Well this is way more subtly weird right now.

Imagine the following code:
```
        Map<String, String> a = new HashMap<>(20);
        a.put("A", "A");

        Map<String, String> b = new HashMap<>();
        b.put("A", "A");
```
Here we have the case that `a.equals(b)` returns `true`. Because that's the contract of Maps!
But when you use `new EqualsBuilder().setTestRecursive(true).append(a, b).isEquals()` then you'll get `false`.
That's because by giving a different initial array size, we'll get a different *internal* (!) structure (bigger internal arrays, etc). 

But that means that right now the EqualsBuilder (even in Java8) does not even adheres to the contract of the Map interface. Because those internal details should not matter. Even if I would not have set an initial size - what if I add 30 elements to one of the 2 maps, but then remove everything but the "A" entry again? It will be equals according to the Map contract, but not for the EqualsBuilder. Looks rather broken to me.

One might argue that that's exactly how all the reflection stuff in commons-lang was designed from the beginning.
Otoh there are already a few places where Maps, Collections, etc get a special treatment. So why not also here?

struberg commented:
> Note that this PR is worse than I thought because we already have a feature to bypass reflection! Use `EqualsBuilder.setBypassReflectionClasses(List<Class<?>>)`.

This is just plain wrong. You cannot use `setBypassReflectionClasses` for e.g. Map without loosing the reflexive comparison of it's entries.
